app/code/Magento/InventoryCatalog/Observer/ProcessSourceItemsObserver.php

```diff
@@ -58,14 +93,57 @@ public function execute(EventObserver $observer)
         }
         /** @var Save $controller */
         $controller = $observer->getEvent()->getController();
+        $productData = $controller->getRequest()->getParam('product', []);
+        $singleSourceData = $productData['quantity_and_stock_status'] ?? [];
 
-        $sources = $controller->getRequest()->getParam('sources', []);
-        $assignedSources = isset($sources['assigned_sources']) && is_array($sources['assigned_sources'])
-            ? $sources['assigned_sources'] : [];
+        if (!$this->isSingleSourceMode->execute()) {
+            $sources = $controller->getRequest()->getParam('sources', []);
```

Comment:
btw have you tested the case when you are in Multi Stock Mode, but did not fill any Source Item data? So, `'sources'` will be empty.
Or is it always has nested `'assigned_sources'` ?

Reply:
In this case "source" is not in the POST array